### ISSUE TITLE
Proposal for a new way to structure Scrapers

### ISSUE THREAD

**Issue by dennyabrain** [tattle-made/tattle-research#10]
I have identified the following problem in the existing structure
1. Large modules that do a lot of tasks intertwined with each other that make for very hard debugging
    1. Debugging issues in one module often involves running the scraper multiple times which leads to downloading data multiple times leading to a very inefficient time consuming debugging
2. Unnecessary dependency on database making it hard to develop and debug locally
3. Lack of historical record of the various steps in scraping. Making it hard to investigate issues and where things went wrong during the course of operation once you discover a bug after weeks/months 

When I reviewed the code and looking at some of the other scrapers we have made since, I've noticed some clear boundaries of separation between tasks which I want to encode in the code. I am just going to propose some Entities and what they should be responsible for. I leave the implementation or ways of organizing the code to the developer (though we can discuss that if you want). As long as we can trace back the Entities mentioned here to some unit in the code, it should be fine.

## Scraper 
We define a Scraper as a unit of code that is run periodically, which finds new articles that need to be scraped, downloads the data from those articles and saves it in Tattle's storage.

Every **Scraper** has a **Domain** and each Domain has an associated Language.

Scrapers have the following entities

### Crawler 
This is the unit responsible for discovering URLs that need to be scraped. In the case of fact checking websites this involves finding the URLs of every newly published article since the last time our scraper run. Time of last scraping can be stored locally in a file or on a remote mongo. Crawlers should be configurable on where should they look for this time.

### ArticleDownloader

For every article discovered in the previous step, we need to download the raw data returned by the server. This needs to be saved in a persistent manner. This will help us have an original copy of the data returned by the fact checking website.

### ArticleParser

This parses the raw data downloaded in the previous step and extracts data of interest, like author name, publication date, article text etc. This entitiy is also responsible for extracting relevant URLs of embedded media like images and videos at the least and tweets, youtube video link and fb links if possible.

### EmbeddedMediaDownloader

The parsed data might container URL to things like files and tweets etc, This entity is responsible for downloading them. Downloading images and videos from their server is an essential functionality task. Its ok to skip tweets, youtube video etc. This entity is also responsible for updating the Parsed data in the previous step with updated information about downloaded files/media.	

### DataFormatter
It takes the parsed data from the previous two steps and structures it in a form that’s suitable for storing in our db. For backwards compatibility reasons we have to store two redundant copies of the data with slightly different structures.

### DataUploader
Uploads the formatted data to remote mongo and the downloaded files to an s3 bucket.

## Additional Notes
Each scraper needs a mode that can be 'local' or 'remote'.
1. in the local mode, the files are downloaded to the local fs and data is stored in a local .json  file
2. in the remote mode, files are uploaded to our s3 bucket and data is stored in our mongoldb

**Comment by tarunima** [comment]
this is a fantastic breakdown. 
Wanted to flag, for article parser- sites could have different data points that could be scraped. Some sites organize content by type of story (politics, health, 'viral'). This would be a good data point to store. We shouldn't be limited by the fields in the existing dataset.

**Comment by tarunima** [comment]
Specifically for Claims Review, we should explore whether the more convenient strategy is to pull it from the sites itself or to pull from the Claims Review API (https://toolbox.google.com/factcheck/apis). This is definitely a long term task and not to be done this sprint.

**Comment by dennyabrain** [comment]
+1 to the point about article parser. each scraper (which corresponds to a website) will most probably have its own article parser because we might be able to extract different fields depending on whats available to scrape. 

Not sure how important this difference is to capture in the code but i feel like most sites have some common fields (date of publication, author name, headline) and some fields unique to that website. the one that comes to mind is that for every article that vishwas news has they also have a dedicated place on the site where they say if the claim is misleading, false or true. so that could be useful to capture. 
so i think it should be easy to look at code of article parser for a scraper and tell what field it will extract. which are standard fields and which are specific to this scraper only. 
This will def also make it very easy to catch exceptions if the structure of the website changes.

**Comment by dennyabrain** [comment]
about your point wrt ClaimsReview, since the claim review data is already on the webpage, it makes sense to do it within the article parser itself no? given the prominece of ClaimsReview, we should define a special Entity (maybe ClaimsReviewParser) within ArticleParser to extract claims review fields.

i was looking at an altnews article and see this clean claims review section in the html :
```
{
  "@context": "http://schema.org",
  "@type": "ClaimReview",
  "datePublished": "2020-10-08",
  "url": "https://www.altnews.in/fact-check-were-rahul-gandhi-and-priyanka-gandhi-cought-laughing-on-way-to-meet-hathras-victim-family/",
  "itemReviewed":
  {
    "@type": "CreativeWork",
    "author":
    {
      "@type": "Person",
      "name": "Anima Sonkar"	  
       	  
    },
    "datePublished": "2016-06-20"
  },
  "claimReviewed": "Rahul and Priyanka Gandhi laughing on their way to Hathras",
  "author":
  {
    "@type": "Organization",
    "name": "Alt News",
        "url" : "https://www.altnews.in"
       },
  "reviewRating":
  {
    "@type": "Rating",
    "ratingValue": "1",
    "bestRating": "5",
    "worstRating": "1",
    "alternateName" : "False"
  }
}
```
pretty straightforward to extract data from. probably an overkill to make API calls to google's endpoints for this?

also my assumption is that all fact checkers who do implement claims review on their site will have such a well structured data in their webpage.

**Comment by duggalsu** [comment]
Hi,

I have implemented the pipeline so that each task can run independently once previous tasks complete. This allows us to 
- Run one task at a time and inspect its output 
- Restart from a specific task in case of failure

Here is the output of each step - 
1. Crawler
    - ../temp_pipeline_output/crawled_urls_<domain>.pkl - pickle file containing a list of crawled urls
    Note: this is needed by the article downloader

2. Article Downloader
    - ../temp_pipeline_output/downloaded_articles_<domain>.pkl - pickle file containing a dict of url and downloaded file path 
```{ url: filepath }```
    Note: this is needed by the parser and data uploader and is only deleted by the data uploader
    - ../data/raw/<domain>/<downloaded articles>.html - downloaded raw html
    - Note: The crawler output file is deleted at this step as future steps are not dependent on it

3. Parser
    - ../temp_pipeline_output/parsed_articles_<domain>.pkl - pickle file containing the following text - "Parse Success!"
   Note: This is only used for failure recovery message when running the article downloader. This is deleted by the data uploader.
    - Writes parsed output to db.
```

```

**Comment by dennyabrain** [comment]
@duggalsu just a clarification regarding "Note: I have not yet fully implemented downloading videos until we decide if it is appropriate to do so"
if the video is hosted on the fact checking website itself then we should download it. if its hosted on something like youtube then we can defer on it.

@tarunima @kruttikanadig can we do a preliminary analysis on seeing what is the norm for fact checkers when they fact check a video, do they have the original video on their own website or do they host it on their youtube account or link to someone else's youtube ?

**Comment by dennyabrain** [comment]
Thank you @duggalsu! what you've written makes sense. eager to see it in action :)

**Comment by tarunima** [comment]
@dennyabrain yes all fact checking sites should contain that structured snippet. 
Will go through a couple of fact checking sites to see what convention on video is.
